File: src/api/schemas/dispatch.py
from pydantic import BaseModel
from uuid import UUID
from datetime import date
from typing import Optional, List, Dict
from src.api.schemas.common import BaseSchema

class WorkerSlot(BaseModel):
    personId: UUID
    name: str
    slotId: UUID

class DaySlot(BaseModel):
    workers: List[WorkerSlot]
    count: int
    required: int
    status: str  # 'fulfilled' | 'partial' | 'shortage'

class ClientDispatchRow(BaseModel):
    clientOrgId: UUID
    clientName: str
    region: Optional[str]
    businessDivision: str
    requiredWorkers: int
    slots: Dict[str, DaySlot]  # ISO date string -> DaySlot

class DispatchSummary(BaseModel):
    totalRequired: int
    totalAssigned: int
    fulfillmentRate: float

class DispatchGridResponse(BaseModel):
    weekStart: date
    weekEnd: date
    clients: List[ClientDispatchRow]
    summary: DispatchSummary

class AvailableWorker(BaseModel):
    personId: UUID
    name: str
    nationality: Optional[str]
    availableFrom: Optional[date]
    skills: List[str] = []

class SimulationChange(BaseModel):
    action: str  # 'add' | 'remove' | 'move'
    personId: UUID
    personName: str
    fromClientId: Optional[UUID] = None
    toClientId: Optional[UUID] = None
    date: date

class SimulationSessionCreate(BaseModel):
    session_name: str
    week_start: date

class SimulationSessionRead(BaseModel):
    sessionId: UUID
    sessionName: str
    weekStart: date
    status: str
    changes: List[SimulationChange] = []

File: types/dispatch.ts
import { BusinessDivision } from './operations';

export interface WorkerSlot {
    personId: string;
    name: string;
    slotId: string;
}

export interface DaySlot {
    workers: WorkerSlot[];
    count: number;
    required: number;
    status: 'fulfilled' | 'partial' | 'shortage';
}

export interface ClientDispatchRow {
    clientOrgId: string;
    clientName: string;
    region: string;
    businessDivision: BusinessDivision;
    requiredWorkers: number;
    slots: Record<string, DaySlot>; // ISO date string -> DaySlot
}

export interface DispatchSummary {
    totalRequired: number;
    totalAssigned: number;
    fulfillmentRate: number;
}

export interface DispatchGridResponse {
    weekStart: string;
    weekEnd: string;
    clients: ClientDispatchRow[];
    summary: DispatchSummary;
}

export interface AvailableWorker {
    personId: string;
    name: string;
    nationality: string;
    availableFrom: string;
    skills: string[];
}

export interface SimulationChange {
    action: 'add' | 'remove' | 'move';
    personId: string;
    personName: string;
    fromClientId?: string;
    toClientId?: string;
    date: string;
}

export interface SimulationSession {
    sessionId: string;
    sessionName: string;
    weekStart: string;
    status: 'draft' | 'reviewing' | 'applied' | 'discarded';
    changes: SimulationChange[];
    createdAt: string;
}
